GitHub Subscribe Organization

benschuster788 opened the pull request:
This is a fix for issue #44 subscribing to an entire Github organization at once.  Subscriptions can occur by providing the full url for the organization or just the organization name.

`/github subscribe org mattermost`
`/github subscribe org https://github.com/mattermost`

The logic for subscribing consists of using Github's built in `ListByOrg` method to access all of the available repositories in an organization.  It then loops through each and subscribes using the previously built `AddSubscription` method.

benschuster788 commented:
@jwilander

hanzei commented:
I would like to keep the organisation as the first of the two parameters of `/github subscribe` for consistency. What do you think of using just `/github subscribe mattermost`?

levb commented on server/utils.go:
```diff
@@ -103,6 +103,20 @@ func decrypt(key []byte, text string) (string, error) {
 	return string(unpadMsg), nil
 }
 
+func parseOwner(full, baseURL string) (string, string) {
```
1/5 would prefer `parseOrg[anization]` as a more intention-revealing name.

benschuster788 replied:
@levb definitely agree with you here

levb commented on server/utils.go:
```diff
@@ -103,6 +103,20 @@ func decrypt(key []byte, text string) (string, error) {
+	if baseURL == "" {
+		baseURL = "https://github.com/"
+	}
+	full = strings.TrimSuffix(strings.TrimSpace(strings.Replace(full, baseURL, "", 1)), "/")
```
0/5 use `url.Parse` and `Path.Split` - more idiomatic?

benschuster788 replied:
@levb this was how it was already being used in parseOwnerAndRepo so just did the same.  Using the url package probably makes the most sense in this case though

benschuster788 commented:
Thanks for the comments all. I wasn't sure how you guys wanted the actual syntax for subscribing to an org.  `/github subscribe [ORG_NAME]` makes a little more sense so I'll update the code to reflect that.

levb commented:
@benschuster788 0/5 - I'd just parse the `git subscribe`'s 1st arg as a [URL/]Path, then if there's only 1 element in the path assume it's the org. Seems unambiguous to me.

benschuster788 commented:
@levb yep that's what I was going to do

benschuster788 commented:
@levb @jasonblais just pushed some updates for this.  Instead of using the new parseOwner method I previously created I modified the parseOwnerAndRepo to handle the case where only an organization is provided.  Then on the command level if it is only an org then it subscribes to the org.

I went ahead and added test coverage for the same method in utils_test.go and updated the `COMMAND_HELP` definition.

I didn't make any changes to the url replacing in that method because the way it is setup is that a full url could be provided like `https://github.com/mattermost` or just the org name `mattermost`.  If we were to do a parse URL while appending the baseURL it could fail despite providing a valid org url (https://github.com/https://github.com/mattermost for examples).  If we want to use url.Parse I can rewrite the method additionally to check if the `full` value is a url already.

jasonblais commented on server/command.go:
```diff
@@ -16,6 +16,7 @@ const COMMAND_HELP = `* |/github connect| - Connect your Mattermost account to y
 * |/github disconnect| - Disconnect your Mattermost account from your GitHub account
 * |/github todo| - Get a list of unread messages and pull requests awaiting your review
 * |/github subscribe list| - Will list the current channel subscriptions
+* |/github subscribe owner [features]| - Subscribe the current channel to all available repositories within organziation and receive notifications about opened pull requests and issues
```
`organziation` > `organization`

benschuster788 replied:
Just fixed!

hanzei commented on server/subscriptions.go:
```diff
@@ -102,6 +98,41 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
+	listOrgOptions := github.RepositoryListByOrgOptions{
+		Type: "all",
+	}
+	repos, _, err := githubClient.Repositories.ListByOrg(ctx, org, &listOrgOptions)
```
The github provides Pages results by default. Hence this will only fetch the first 100 (I think) repositories. See https://github.com/google/go-github#pagination for more infos.

hanzei commented on server/subscriptions.go:
```diff
@@ -102,6 +98,41 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
+			Repository: fmt.Sprintf("%s/%s", org, repo.GetFullName()),
+		}
+
+		if err := p.AddSubscription(fmt.Sprintf("%s/%s", org, repo), sub); err != nil {
```
Maybe log the error?

benschuster788 replied:
That’s a good idea, I’ll add that in

levb commented:
@benschuster788 When merging the ticket I accidentally overlooked that there were 2 comments from @hanzei still unresolved. Do you mind fixing those? Apologies for the confusion,

benschuster788 commented:
@levb @hanzei just created a new PR #56 with a commit that paginates through the repos in a organization since the limit returned by Github is just 30 repos per response.
